style-guide.md

```diff
@@ -0,0 +1,165 @@
+# edX JavaScript Conventions
+
+# TODO
+
+I'm looking for widespread agreement on everything contained in this document!
+In particular, I'd love feedback on the [Goals](#goals), [Style](#style), and
+[Conventions](#conventions) sections.
+
+Here are a few things that would be awesome to have:
+- [ ] define and link to a jshint/jscs/eslint config
+- [ ] clarify path towards ES2015
+
+## Motivation
+
+Code conventions are important.  It helps us intuitively approach new code,
+revisit old code, and share code with others.
+
+## Goals
+
+We hope to write JavaScript which is:
+
+1. Legible
+2. Friendly
+3. Predictable
+4. Semantic
+
+## About
+
+This guide is split up into two sections - [Style](#style) and
+[Conventions](#conventions).  The style section addresses aesthetic topics,
+while the conventions section addresses topics that have more semantic
+significance.
+
+## Note: ECMAScript Version
+
+Currently, edX only supports ES5.  In order to implement modules, we use AMD by
+way of [RequireJS](http://requirejs.org/).  We are moving towards ES2015
+however, and this guide will be updated when we get there.
+
+## Style
+### Indentation
+
+All JavaScript should be indented with 4 spaces.  edX uses this standard
+throughout our Python, HTML, SASS/CSS, and JavaScript code.
+
+e.g.
+
+```javascript
+function sayHello (name) {
+    return 'Hello, ' + name;
+}
+```
+
+### Semicolons
+
+Use them!
+
+### Line Length
+
+Break lines at `120` columns.
+
+### Strings
+
+Use single quotes for string literals, unless the string itself contains single
+quotes.
+
+e.g.
+
+```javascript
+var edX = 'edX',
+    thisGuide = "edX's JavaScript Style Guide";
+```
+
+### Comments
+
+If code does not read clearly on its own, consider breaking it up into smaller
+pieces. If things are still not clear, add a descriptive comment in the form of
+full-sentence(s).
+
+### Variables
+
+We like to write our code to reflect the fact that JavaScript hoists `var`
+declarations to the top-level environment of the current function. Therefore
+declare all variables for a particular function at the start of the function.
+
+e.g.
+
+```javascript
+var enrollUserInCourse = function (username, courseId) {
+    var userModel,
+        courseModel;
```

Comment:
@dan-f I have been a fan of one `var` at the top of function. But apart from hoisting and less typing there is no reason to use it. What we have learned from other languages is that variable declaration should be as close to it's use as possible, it makes the reading much easier. With ES6 when we will be writing `let` this pattern will go away I think. getify (you don't know js) has a very different opinion [https://twitter.com/getify/status/726917076812599296](https://twitter.com/getify/status/726917076812599296).

PS. while debugging all statements and expressions with one `var` evaluate at the same time.